### ISSUE TITLE
[2.6.0] TBR sometimes not shown correct in start screen Insight pump

### ISSUE THREAD

**Issue by Chris99621** [MilosKozak/AndroidAPS#2484]
Sometimes AAPS dont show in the start screen the correct TBR (under BAS) like there, but the TBR in the pump ist set: 
![buggedScreenshot_20200306-130315_AndroidAPS](https://user-images.githubusercontent.com/61082188/76088233-f27aa800-5fb7-11ea-8c9c-d44f67a9a8f6.jpg)
![buggedScreenshot_20200306-130330_AndroidAPS](https://user-images.githubusercontent.com/61082188/76088238-f3abd500-5fb7-11ea-9235-5cd98af4c2c9.jpg)
sometimes it is correct (i mean the blue font)
![okScreenshot_20200306-141344_AndroidAPS](https://user-images.githubusercontent.com/61082188/76088322-13db9400-5fb8-11ea-84b4-72409d18bbaa.jpg)

Running on AAPS 2.6.0 on Galaxy S10+ with Insight and Libre 2
Logs (Date:06.03.2020, Time: see in screenshots): 
[AndroidAPS.log](https://github.com/MilosKozak/AndroidAPS/files/4298574/AndroidAPS.log)
[Log Export.zip](https://github.com/MilosKozak/AndroidAPS/files/4298588/Log.Export.zip)

**Comment by chrisbev** [comment]
Same for me. Insight and 2.6. Problem also in 2.5.1 but the work around for 2.5.1 to not use absolute values and use disabled NS sync doesn't work for me in 2.6. Nor disabling all NS. Samsung A3 2017 Android  8.0.0 and MiaoMiao Libre Glimp. Looking for trigger points as problem only sometimes. Seen when AAPS disconnect pump. Bg falling through target level. BG falling and hitting Automation level set to raise Target. Noticed points have been pump and loop tabs at 0% but graph at 100%, I think Home BAS: also at 100% level. Haven't noticed any situations where graph is less than pump but not looked deliberately for them as it is where I have seen graph displaying insulin when I would expect none i.e. low BG and in fact pump was 0%. If a solution could be found that would be good as missing Autotune since 2.5.1!

**Comment by MilosKozak** [comment]
do you have upload as absolute TBR disabled?

**Comment by Chris99621** [comment]
i did these are my settings
![1](https://user-images.githubusercontent.com/61082188/76122750-1f01e480-5ff7-11ea-89e6-25d70655d90d.jpg)
i just recognized that these settings dont work, the upload ist working but the device ist not loading
![2](https://user-images.githubusercontent.com/61082188/76122812-4ce72900-5ff7-11ea-8ec9-c00d417c0323.jpg)

**Comment by chrisbev** [comment]
Hi yes screenshot attached, I did try with all NS off now trying again with
disabled sync to see what gets to NS as @woodmasterslick asked.

**Comment by overfrenk** [comment]
Same problem too. In 2.5.1 not have this problem!
My setting is NS:upload only ON and absolute TBR OFF.

**Comment by chrisbev** [comment]
@overfrenk I don't know if it is everyone with Insight who has this problem or not. I thought the problem was know as there was a workaround for 2.5.1 so just waiting but it seems not. Since I've been trying to get more data to help the devs, just lately I haven't seen the problem. Now as it only sometimes occurs for me it may bite back. A thought has gone through my mind perhaps you could try? I am wondering if turning the NS switches and putting them back somehow 'corrected' things? I switched the 2 NS ones and the absolute one. Probably switched then came out of settings then went back? Just a thought it would good if it worked and even better to get autotune back.

**Comment by chrisbev** [comment]
Hi @overfrenk I 'captured' an 'event' early 
this morning so please ignore my sugestion above. I'll attach pictures and log a bit later.

**Comment by chrisbev** [comment]
Situation: Woke at 6:30 BGs falling HomeScreen graph shows 100% but pump tab 20%. Bathroom visit and BG rose prompting some insulin. Inspecting NS graph the 20% (NS -80%) is shown but from atteached not on the HSgraph. Secondly just before 8:00 HSgraph shows anout 30mins of 100% but these are not in NS nor in the pump data logs.AndroidAPS._2020-03-08_00-00-00_.1.zip](https://github.com/MilosKozak/AndroidAPS/files/4303081/AndroidAPS._2020-03-08_00-00-00_.1.zip)
![Screenshot_20200308-063103_AndroidAPS](https://user-images.githubusercontent.com/43484095/76161212-02fe6000-6129-11ea-8991-5f28046f911b.jpg)
![Screenshot_20200308-063116_AndroidAPS](https://user-images.githubusercontent.com/43484095/76161219-1d383e00-6129-11ea-9301-e58039ef7047.jpg)
![Screenshot_20200308-064533_Samsung Internet](https://user-images.githubusercontent.com/43484095/76161233-32ad6800-6129-11ea-96dd-f374cf1b853d.jpg)
![Screenshot_20200308-080357_Samsung Internet](https://user-images.githubusercontent.com/43484095/76161238-435dde00-6129-11ea-99d5-ad34b5da0cec.jpg)
![Screenshot_20200308-080415_AndroidAPS](https://user-images.githubusercontent.com/43484095/76161242-4bb61900-6129-11ea-8879-9efe299cb939.jpg)

**Comment by chrisbev** [comment]
If it helps have also seen times when  above target with BAS: and graph at 100% but pump at well over 100%.

**Comment by reijoahola-nightscout** [comment]
I moved from Combo -> Insight yesterday and noticed also this kind of problems. I have used setting "NS upload only". But when I removed that setting and syncing now back from NS, the main window basal and Treatments/Temp basal follow what is in the pump. May be syncing back from NS corrects the situation.

**Comment by reijoahola-nightscout** [comment]
What I also noticed that now with Insight AAPS removes TBR directly from some higher or lower value into 100% (normal) basal, without using 90% or 110% steps. Is this changed intentionally because the beep from pump can be sileced when using Insight ?

**Comment by reijoahola-nightscout** [comment]
This is not only display problem in the main window but this also affects basal IOB calculation because Treatments/Temp basal lines were not accrding to pump values, at lest in my situation.

**Comment by artirgeci** [comment]
> This is not only display problem in the main window but this also affects basal IOB calculation because Treatments/Temp basal lines were not accrding to pump values, at lest in my situation.

I have been observing the same situation. AC Insight.

**Comment by artirgeci** [comment]
The last 24h I disabled "NS upload only" on NSClient, and the information has been correct between the Home screen and the pump. For me the discrepancies were almost all the time when any TBR was active. I will keep you updated if I notice anything in the coming days. 
![Screenshot_20200314-224523](https://user-images.githubusercontent.com/23337996/76690957-b4c7f000-6645-11ea-8efb-1755b417f931.jpg)
![Screenshot_20200314-224528](https://user-images.githubusercontent.com/23337996/76690958-b691b380-6645-11ea-9dad-b83e7eb868b6.jpg)
![Screenshot_20200314-223424](https://user-images.githubusercontent.com/23337996/76690959-b7c2e080-6645-11ea-8829-db8363e72f44.jpg)

**Comment by reijoahola-nightscout** [comment]
When using "NS upload only" =disabled, I have also not noticed this problem anymore. Just worried if problem still exist when network connection is temporarily off (Airplane mode etc.). Better to test this in some phase today.

**Comment by Chris99621** [comment]
testet the solution from @artirgeci to deaktivate NSupload only like in the screenshot above and the whole iob kalkulation changed. Time ca. 21:08. I think the bug is bigger than expected.

Log: 
[AndroidAPS.log](https://github.com/MilosKozak/AndroidAPS/files/4351173/AndroidAPS.log)

**Comment by artirgeci** [comment]
> testet the solution from @artirgeci to deaktivate NSupload only like in the screenshot above and the whole iob kalkulation changed. Time ca. 21:08. I think the bug is bigger than expected.
> 
> Log:
> [AndroidAPS.log](https://github.com/MilosKozak/AndroidAPS/files/4351173/AndroidAPS.log)

@Chris99621  Since 5 days now with "NS upload only" =disabled, everything is working fine for me. Is it working for you also? Sorry I didn’t go through your log file.

**Comment by reijoahola-nightscout** [comment]
I have also now used "NS upload only" =disabled for several days without this problem. However, I do not consider this problem yet solved because I am worried if this problem starts again when network connection happens to be offline, e.g. Airplane mode or network just stops working.

**Comment by Chris99621** [comment]
@artirgeci yes it works with NS upload only = disabled for me. But if i pause the NS connection it wont work

**Comment by artirgeci** [comment]
> @artirgeci yes it works with NS upload only = disabled for me. But if i pause the NS connection it wont work

@Chris99621 that's interesting I'll try that. Although i did a test putting my phone in airplane mode, it still worked.

**Comment by Chris99621** [comment]
> 
> 
> > @artirgeci yes it works with NS upload only = disabled for me. But if i pause the NS connection it wont work
> 
> @Chris99621 that's interesting I'll try that. Although i did a test putting my phone in airplane mode, it still worked.

ok in my settings i disabeld NS upload only but i paused the NS connection, but if i started it the iob kalkulation changed like i posted above

**Comment by artirgeci** [comment]
> > > @artirgeci yes it works with NS upload only = disabled for me. But if i pause the NS connection it wont work
> > 
> > 
> > @Chris99621 that's interesting I'll try that. Although i did a test putting my phone in airplane mode, it still worked.
> 
> ok in my settings i disabeld NS upload only but i paused the NS connection, but if i started it the iob kalkulation changed like i posted above

@Chris99621 ah I see, thanks. I'll try it in the coming days.

**Comment by Suchiman** [comment]
I've analysed the logs and read a bit through the code, when it works the events happen in this order
```
14:53:44.695 [Thread-12032] D/DATABASE: [DatabaseHelper.roundDateToSec():361]: Rounding 1584885224690 to 1584885224000
14:53:44.705 [Thread-12032] D/DATABASE: [DatabaseHelper.createOrUpdate():908]: TEMPBASAL: New record from: PUMP TemporaryBasal{date=1584885224000, date=22.03.20 14:53, isValid=true, pumpId=1577118934874, _id=null, percentRate=0, absoluteRate=0.0, durationInMinutes=0, isAbsolute=false, isFakeExtended=false, netExtendedRate=0.0}
...
14:53:44.713 [Thread-12032] D/DATABASE: [DatabaseHelper.roundDateToSec():361]: Rounding 1584885225676 to 1584885225000
14:53:44.721 [Thread-12032] D/DATABASE: [DatabaseHelper.createOrUpdate():908]: TEMPBASAL: New record from: PUMP TemporaryBasal{date=1584885225000, date=22.03.20 14:53, isValid=true, pumpId=1577118934873, _id=null, percentRate=0, absoluteRate=0.0, durationInMinutes=90, isAbsolute=false, isFakeExtended=false, netExtendedRate=0.0}
```
and when it doesn't work, this happens
```
15:08:49.370 [Thread-12155] D/DATABASE: [DatabaseHelper.roundDateToSec():361]: Rounding 1584886128349 to 1584886128000
15:08:49.381 [Thread-12155] D/DATABASE: [DatabaseHelper.createOrUpdate():908]: TEMPBASAL: New record from: PUMP TemporaryBasal{date=1584886128000, date=22.03.20 15:08, isValid=true, pumpId=1577118934875, _id=null, percentRate=0, absoluteRate=0.0, durationInMinutes=120, isAbsolute=false, isFakeExtended=false, netExtendedRate=0.0}
...
15:08:49.391 [Thread-12155] D/DATABASE: [DatabaseHelper.roundDateToSec():361]: Rounding 1584886128362 to 1584886128000
15:08:49.393 [Thread-12155] D/DATABASE: [DatabaseHelper.createOrUpdate():896]: TEMPBASAL: Updated record with Pump Data : PUMP TemporaryBasal{date=1584886128000, date=22.03.20 15:08, isValid=true, pumpId=1577118934876, _id=null, percentRate=0, absoluteRate=0.0, durationInMinutes=0, isAbsolute=false, isFakeExtended=false, netExtendedRate=0.0}
```
So when it doesn't work, it appears that the cancel TBR record (duration = 0) comes after the set TBR record and updates the set TBR record to be the cancel TBR record.

**Comment by reijoahola-nightscout** [comment]
Any news in solving this issue ? I have used a temporary solution (NS upload only setting disabled) but it cannot be considered as the final solution. The problen is not only how basal is displayed but it causes also error in  basal IOB calculation what can be dangerous. Have somebody tested what happens if NS network connection happens to be down temporarily ?

**Comment by artirgeci** [comment]
> Any news in solving this issue ? I have used a temporary solution (NS upload only setting disabled) but it cannot be considered as the final solution. The problem is not only how basal is displayed but it can cause also error in basal IOB calculation what can be dangerous. Have somebody tested what happens if NS network connection happens to be down temporarily ?

I'm still using the temp solution as you. I did not test what happens when offline, but it happened to me during the night to lose the connection, as I can remember it went to my normal basal rate. As it was during the night I really did not analyze if it had any other impact, I just restarted the client to come back online again. Hopefully there will be a solution in the next releases.

**Comment by chrisbev** [comment]
Hi R..,
With 2.6.1 I had a temp solution with the switches but with 2.6.1.2  I have
not and it is SO annoying for me. When going low the blue graph (& BAS)
would be a quick check but it is  sometimes (and seems for me to be quite
often) wrong so have to check the Insight tab and just to be sure the pump!
There seems a couple of suggested  'fixes' for this issue and #2249 #2548
but I can't seem to build an apk and run closed loop to be able to try so
waiting.
It looks like it is a sequence of timing issues due to the one second
resoultion that only a few selected people get! Maybe due to use or
settings or body?
Rgds, Chris

**Comment by Elektrocat** [comment]
Is it possible that this problem has something to do with the TBR-emulation set to "on" what does not really make sense using SMB? I have a V3 insight and at the moment this problem is gone with 2.6.1.4.
But users of older pump-FW still have that big issue.

**Comment by reijoahola-nightscout** [comment]
Could somebody test this issue with the new release 2.6.7-beta1. I compiled and installed it but because I can use it only in my test phone (virtual pump and NS not connected) I cannot verify has the problem disappeared. Also because the release is dev version it allows only open loop mode so it can somewhat limit how reliable the test can be. I am interested to hear how the new release works with Insight pump V2 and when the setting "NS upload only (disabled sync)" is ON.

**Comment by Elektrocat** [comment]
It works with V2. (testet by a friend of mine) He says he CAN use closed loop as well.

**Comment by reijoahola-nightscout** [comment]
Hi Electrocat
> It works with V2. (testet by a friend of mine) He says he CAN use closed loop as well.
Could you ask your friend if he had this issue with the earlier version because it looks like
all Insight users do not experience this problem.

That Closed Loop problem what I mentioned was something related to this dev/Beta version.
When I turned Engineering Mode on  then I could also use Closed Loop.
My test system is not enough realistic (Insight pump not installed) that I cannot say anything does this now work or not.
I do not want yet to install this beta version into my main AAPS phone.

**Comment by Chris99621** [comment]
@reijoahola-nightscout i testet it with 2.6.7-beta1 for 2 days and NS uplopad only. Can´t see the bug. It works now as it should.

**Comment by artirgeci** [comment]
Using for about 24h now. All looks good. Insight v2.

**Comment by chrisbev** [comment]
Hi evaluators are you seeing any aaps/ns differences like #2700 pictures where seems to be when there are tbr and extendeds i.e. exceeding 250%? (For me difference didn't happen very often may be one a day?) I seemed to have had a problem with building beta1 in that it reported as master and told to try dev. Sorry a different Chris!

**Comment by reijoahola-nightscout** [comment]
I have also build 2.6.7-beta1 but only used in my test system (only virtual pump) because I am not sure have I compiled correctly. It have some peculiarities and would like to hear your comments. 1st thing was that when opened it continously reported being dev version even if About-status showed 2.6.7-beta1 without dev-texts and so on. So it refused to use Closed Loop mode. When I turned engineering_mode on then Closed Loop was possible. 2nd problem was that when using Low Glucose Suspend mode it forced max_iob = 0. Should it work like that ? In the normal Closed Loop mode it seemed to work better when AMA was used but then I noticed in SMB mode that max_iob was not 4 that I had set into parameter but it was always 5 independent from the setting. Have I messed something or are these bugs ?

**Comment by TebbeUbben** [comment]
This bug occurs when a new TBR is set to override the current one and the pump logs the end event at the same second as the start event. Due to AndroidAPS not supporting two events having the same timestamp, the end event takes precedence. In 2.7.0 we add some artificial offset to end events to avoid this issue.

**Comment by Chris99621** [comment]
@TebbeUbben i testet it since the extra insight branch and it works fine. I think this can be closed as soon as 2.7.0 is released.

**Comment by reijoahola-nightscout** [comment]
> This bug occurs when a new TBR is set to override ....
I am now testing with 2.7.0-rc2 by using Insight v2 pump. Sync from NS is disabled. It is the setting where I had earlier this problem. After one day testing I have not found the problem anymore but I will continue the test. This is my main platform with real G6 sensor and real Insight pump. I am using local profiles but CGM, Treatments etc. are updated into NS (but not synced back).
